File: scripts/fetch-with-session.js
const puppeteer = require('puppeteer');
const fs = require('fs-extra');
const path = require('path');

async function fetchWithSession(categoryId = '252', finished = true) {
  console.log('🚀 Launching browser with session persistence...');
  
  const browser = await puppeteer.launch({
    headless: false, // НЕ headless для обхода Cloudflare
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--no-first-run',
      '--disable-gpu',
      '--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    ]
  });
  
  const page = await browser.newPage();
  
  // Настройка viewport
  await page.setViewport({ width: 1920, height: 1080 });
  
  // Дополнительные заголовки
  await page.setExtraHTTPHeaders({
    'Accept-Language': 'ru-RU,ru;q=0.9,en;q=0.8',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'Cache-Control': 'no-cache',
    'Pragma': 'no-cache'
  });
  
  const url = `https://meshok.net/good/${categoryId}${finished ? '?opt=2' : ''}`;
  
  console.log(`📄 Fetching: ${url}`);
  console.log('⚠️  Browser will open in GUI mode - you may need to manually pass Cloudflare challenge');
  console.log('⏳ Waiting 30 seconds for manual intervention...');
  
  try {
    await page.goto(url, {
      waitUntil: 'domcontentloaded',
      timeout: 60000
    });
    
    // Ждем 30 секунд для ручного прохождения Cloudflare
    await new Promise(resolve => setTimeout(resolve, 30000));
    
    const content = await page.content();
    
    // Сохраняем результат
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `session_good${categoryId}_opt${finished ? '2' : '1'}_${timestamp}.html`;
    const filepath = path.join(__dirname, '../data', filename);
    
    await fs.ensureDir(path.join(__dirname, '../data'));
    await fs.writeFile(filepath, content, 'utf-8');
    
    console.log(`✅ Saved to: ${filename}`);
    console.log(`📊 Size: ${(content.length / 1024).toFixed(2)} KB`);
    
    // Анализ содержимого
    const title = await page.title();
    console.log(`📋 Page title: ${title}`);
    
    // Поиск ссылок на лоты
    const lotLinks = await page.$$eval('a[href*="/item/"]', links => links.length);
    console.log(`🔗 Found ${lotLinks} item links`);
    
    if (lotLinks > 0) {
      console.log('🎉 Successfully obtained auction data!');
      
      // Показываем первые несколько ссылок
      const firstLinks = await page.$$eval('a[href*="/item/"]', links => 
        links.slice(0, 5).map(link => link.href)
      );
      console.log('📋 First 5 item links:');
      firstLinks.forEach((link, index) => {
        console.log(`   ${index + 1}. ${link}`);
      });
    } else {
      console.log('⚠️  No auction links found');
    }
    
  } catch (error) {
    console.error('❌ Error:', error.message);
  } finally {
    console.log('⏳ Browser will stay open for 10 more seconds...');
    await new Promise(resolve => setTimeout(resolve, 10000));
    await browser.close();
    console.log('🏁 Browser closed');
  }
}

// Запуск
const categoryId = process.argv[2] || '252';
const finished = process.argv[3] !== 'false';
fetchWithSession(categoryId, finished);

File: scripts/fetch-with-session.py
#!/usr/bin/env python3

import sys
import os
import time
import json
from datetime import datetime
import urllib.request
import urllib.parse
from urllib.error import URLError, HTTPError
import subprocess

def fetch_with_session(category_id='252', finished=True):
    print('🍪 Using session-based approach...')
    
    url = f"https://meshok.net/good/{category_id}{'?opt=2' if finished else ''}"
    print(f"📄 Fetching: {url}")
    
    try:
        # Сначала получаем главную страницу для получения cookies
        print('⏳ Getting main page for session...')
        
        main_req = urllib.request.Request(
            'https://meshok.net/',
            headers={
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
                'Accept-Language': 'en-US,en;q=0.9',
                'Accept-Encoding': 'gzip, deflate, br',
                'Cache-Control': 'no-cache',
                'Pragma': 'no-cache',
                'Sec-Fetch-Dest': 'document',
                'Sec-Fetch-Mode': 'navigate',
                'Sec-Fetch-Site': 'none',
                'Sec-Fetch-User': '?1',
                'Upgrade-Insecure-Requests': '1',
                'DNT': '1'
            }
        )
        
        # Создаем cookie jar для сохранения сессии
        import http.cookiejar
        cookie_jar = http.cookiejar.CookieJar()
        opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(cookie_jar))
        
        try:
            with opener.open(main_req, timeout=30) as response:
                main_content = response.read().decode('utf-8')
                print(f'✅ Main page loaded: {len(main_content) / 1024:.2f} KB')
                
                # Ждем немного между запросами
                time.sleep(2)
                
                # Теперь делаем запрос к целевой странице с cookies
                print('⏳ Making request to target page with session...')
                
                target_req = urllib.request.Request(
                    url,
                    headers={
                        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
                        'Accept-Language': 'en-US,en;q=0.9',
                        'Accept-Encoding': 'gzip, deflate, br',
                        'Cache-Control': 'no-cache',
                        'Pragma': 'no-cache',
                        'Sec-Fetch-Dest': 'document',
                        'Sec-Fetch-Mode': 'navigate',
                        'Sec-Fetch-Site': 'same-origin',
                        'Sec-Fetch-User': '?1',
                        'Upgrade-Insecure-Requests': '1',
                        'DNT': '1',
                        'Referer': 'https://meshok.net/'
                    }
                )
                
                with opener.open(target_req, timeout=30) as response:
                    content = response.read().decode('utf-8')
                    
                    print(f'📊 Status code: {response.status}')
                    print(f'📊 Response size: {len(content) / 1024:.2f} KB')
                    
                    # Проверяем на Cloudflare
                    if 'Just a moment' in content or 'Один момент' in content:
                        print('⚠️  Cloudflare challenge detected')
                    else:
                        print('✅ No Cloudflare challenge detected')
                    
                    # Сохраняем результат
                    timestamp = datetime.now().strftime('%Y-%m-%dT%H-%M-%S-%fZ')
                    filename = f"session_good{category_id}_opt{'2' if finished else '1'}_{timestamp}.html"
                    filepath = os.path.join('data', filename)
                    
                    os.makedirs('data', exist_ok=True)
                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(content)
                    
                    print(f'✅ Saved to: {filename}')
                    
                    # Простой анализ содержимого
                    if '<title>' in content:
                        title_start = content.find('<title>') + 7
                        title_end = content.find('</title>')
                        if title_end > title_start:
                            title = content[title_start:title_end]
                            print(f'📋 Page title: {title}')
                    
                    # Поиск ссылок на лоты
                    item_links = content.count('href="/item/')
                    print(f'🔗 Item links found: {item_links}')
                    
                    if item_links > 0:
                        print('🎉 Successfully obtained auction data with session!')
                        
                        # Извлекаем первые несколько ссылок
                        import re
                        link_matches = re.findall(r'href="(/item/[^"]*)"', content)
                        print('📋 First 5 item links:')
                        for i, link in enumerate(link_matches[:5]):
                            print(f'   {i + 1}. https://meshok.net{link}')
                    else:
                        print('⚠️  No auction links found')
                    
                    # Поиск цен
                    price_matches = re.findall(r'[0-9,]+[ ]*₽|[0-9,]+[ ]*руб', content)
                    if price_matches:
                        print(f'💰 Prices found: {len(price_matches)}')
                        print('📋 Sample prices:')
                        for i, price in enumerate(price_matches[:3]):
                            print(f'   {i + 1}. {price}')
                    
                    # Поиск таблиц
                    table_count = content.count('<table')
                    print(f'📊 Tables found: {table_count}')
                    
                    # Поиск форм
                    form_count = content.count('<form')
                    print(f'📝 Forms found: {form_count}')
                    
                    # Поиск JSON данных
                    json_matches = re.findall(r'\{[^{}]*"[^"]*"[^{}]*\}', content)
                    if json_matches:
                        print(f'📜 JSON data found: {len(json_matches)} matches')
                        print('📋 Sample JSON:')
                        for i, json_data in enumerate(json_matches[:2]):
                            print(f'   {i + 1}. {json_data[:100]}...')
                    else:
                        print('📜 No JSON data found')
                        
        except HTTPError as e:
            print(f'❌ HTTP Error: {e.code} - {e.reason}')
            if e.code == 403:
                print('💡 403 Forbidden - Cloudflare is blocking requests')
                print('💡 Try using a different approach or proxy')
        except URLError as e:
            print(f'❌ URL Error: {e.reason}')
        except Exception as e:
            print(f'❌ Error: {e}')
            
    except Exception as e:
        print(f'❌ Error: {e}')

if __name__ == '__main__':
    category_id = sys.argv[1] if len(sys.argv) > 1 else '252'
    finished = sys.argv[2] != 'false' if len(sys.argv) > 2 else True
    fetch_with_session(category_id, finished)
